### PR TITLE
stargate: split x/evm/ (minus types/)

### PR DESCRIPTION
**Pull request by fedekunze**
- stargate: split proto and base files
- stargate: split app/ + codec/
- stargate: split types/
- stargate: split crypto/
- stargate: split x/evm/types/
- stargate: split x/evm/ (minus types/)

### CODE REVIEW

File: x/evm/client/rest/rest.go [hunk 1]
```diff
@@ -1,90 +1,89 @@
...
 		vars := mux.Vars(r)
 		hashHexStr := vars["hash"]
 
-		cliCtx, ok := rest.ParseQueryHeightOrReturnBadRequest(w, cliCtx, r)
+		clientCtx, ok := rest.ParseQueryHeightOrReturnBadRequest(w, clientCtx, r)
 		if !ok {
 			return
 		}
 
 		ethHashPrefix := "0x"
```
Comment (freddyli7) [inline x/evm/client/rest/rest.go:44]: do we have somewhere to predefine all const like address length, sig length and prefix? maybe that is where to store them, easy to manage and maintain.

File: x/evm/module.go [hunk 2]
```diff
@@ -18,71 +22,87 @@ import (
...
+// Name returns the evm module's name.
 func (AppModuleBasic) Name() string {
 	return types.ModuleName
 }
 
-// RegisterCodec registers types for module
-func (AppModuleBasic) RegisterCodec(cdc *codec.Codec) {
-	types.RegisterCodec(cdc)
+// RegisterLegacyAminoCodec performs a no-op as the evm module doesn't support amino.
+func (AppModuleBasic) RegisterLegacyAminoCodec(_ *codec.LegacyAmino) {
```
Comment (freddyli7) [inline x/evm/module.go:39]: should we remove this func? since it's doing nothing and return nothing as well.

File: x/evm/module.go [hunk 3]
```diff
@@ -18,71 +22,87 @@ import (
...
 	}
 
 	return genesisState.Validate()
 }
 
-// RegisterRESTRoutes Registers rest routes
-func (AppModuleBasic) RegisterRESTRoutes(ctx context.CLIContext, rtr *mux.Router) {
+// RegisterRESTRoutes performs a no-op as the EVM module doesn't expose REST
+// endpoints
+func (AppModuleBasic) RegisterRESTRoutes(_ client.Context, _ *mux.Router) {
```
Comment (freddyli7) [inline x/evm/module.go:60]: same as above

File: x/evm/module.go [hunk 4]
```diff
@@ -18,71 +22,87 @@ import (
...
+func (AppModuleBasic) RegisterGRPCGatewayRoutes(clientCtx client.Context, mux *runtime.ServeMux) {
+	if err := types.RegisterQueryHandlerClient(context.Background(), mux, types.NewQueryClient(clientCtx)); err != nil {
+		panic(err)
+	}
 }
 
-// GetTxCmd Gets the root tx command of this module
-func (AppModuleBasic) GetTxCmd(cdc *codec.Codec) *cobra.Command {
+// GetTxCmd returns nil as the evm module doesn't support transactions through the CLI.
+func (AppModuleBasic) GetTxCmd() *cobra.Command {
```
Comment (freddyli7) [inline x/evm/module.go:71]: this one returns nil but no longer supported, is it safe to remove? otherwise it might be called and cause nil pointer error

**Comment by freddyli7** [comment]
made some comments but not sure if removing those unsupported funcs are safe.

File: x/evm/handler_test.go [hunk 5]
```diff
@@ -264,31 +215,31 @@ func (suite *EvmTestSuite) TestQueryTxLogs() {
...
-	suite.codec.MustUnmarshalJSON(res, &txLogs)
+	// var txLogs types.QueryETHLogs
+	// suite.codec.MustUnmarshalJSON(res, &txLogs)
 
 	// amino decodes an empty byte array as nil, whereas JSON decodes it as []byte{} causing a discrepancy
-	resultData.Logs[0].Data = []byte{}
-	suite.Require().Equal(txLogs.Logs[0], resultData.Logs[0])
+	// txResponse.TxLogs.Logs[0].Data = []byte{}
+	// suite.Require().Equal(txLogs.Logs[0], txResponse.TxLogs.Logs[0])
 }
```
Comment (noot) [inline x/evm/handler_test.go:243]: why are these commented out?

File: x/evm/keeper/abci.go [hunk 6]
```diff
@@ -11,7 +11,7 @@ import (
 	ethtypes "github.com/ethereum/go-ethereum/core/types"
 )
 
-// BeginBlock sets the block hash -> block height map for the previous block height
+// BeginBlock sets the block height -> header hash map for the previous block height
```
Comment (noot) [inline x/evm/keeper/abci.go:14]: if there's no hash -> height map how does `eth_getBlockByHash` work?

File: x/evm/keeper/keeper.go [hunk 7]
```diff
@@ -66,31 +68,6 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
 }
 
-// ----------------------------------------------------------------------------
-// Block hash mapping functions
-// Required by Web3 API.
-//  TODO: remove once tendermint support block queries by hash.
-// ----------------------------------------------------------------------------
-
-// GetBlockHash gets block height from block consensus hash
-func (k Keeper) GetBlockHash(ctx sdk.Context, hash []byte) (int64, bool) {
```
Comment (noot) [inline x/evm/keeper/keeper.go:76]: why are these removed? are they moved?

File: x/evm/module.go (hunk 4 above)
Comment (fedekunze) [reply x/evm/module.go:71]: it needs to fulfil the module interface so it is ok to return nil

File: x/evm/keeper/abci.go (hunk 6 above)
Comment (fedekunze) [reply x/evm/keeper/abci.go:14]: I added it to the tendermint client so it's directly supported by the node 🙂

Comment (fedekunze) [reply x/evm/keeper/abci.go:14]: good question btw

File: x/evm/keeper/keeper.go (hunk 7 above)
Comment (fedekunze) [reply x/evm/keeper/keeper.go:76]: yup, not longer necessary

File: x/evm/module.go (hunk 2 above)
Comment (fedekunze) [reply x/evm/module.go:39]: it's required by the module interface

File: x/evm/module.go (hunk 3 above)
Comment (fedekunze) [reply x/evm/module.go:60]: it's required by the module interface

File: x/evm/client/cli/query.go [hunk 8]
```diff
@@ -1,88 +1,105 @@
...
-func GetCmdGetStorageAt(queryRoute string, cdc *codec.Codec) *cobra.Command {
-	return &cobra.Command{
-		Use:   "storage [account] [key]",
-		Short: "Gets storage for an account at a given key",
+// GetStorageCmd queries a key in an accounts storage
+func GetStorageCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "storage [address] [key]",
+		Short: "Gets storage for an account with a given key and height",
+		Long:  "Gets storage for an account with a given key and height. If the height is not provided, it will use the latest height from context.",
```
Comment (araskachoi) [inline x/evm/client/cli/query.go:35]: should we change `account` to `address` in the description as well?
this address will be the hex ethereum address correct? was this previously the cosmos account address passed?

File: x/evm/client/cli/query.go [hunk 9]
```diff
@@ -1,88 +1,105 @@
...
-// GetCmdGetCode queries the code field of a given address
-func GetCmdGetCode(queryRoute string, cdc *codec.Codec) *cobra.Command {
-	return &cobra.Command{
-		Use:   "code [account]",
+// GetCodeCmd queries the code field of a given address
+func GetCodeCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "code [address]",
 		Short: "Gets code from an account",
+		Long:  "Gets code from an account. If the height is not provided, it will use the latest height from context.",
```
Comment (araskachoi) [inline x/evm/client/cli/query.go:75]: switch wording to address here too?

File: x/evm/client/rest/rest.go [hunk 10]
```diff
@@ -1,90 +1,89 @@
...
 		return nil, err
 	}
-	node, err := cliCtx.GetNode()
+	node, err := clientCtx.GetNode()
 	if err != nil {
 		return nil, err
 	}
-	tx, err := node.Tx(hash, false)
+
+	tx, err := node.Tx(context.Background(), hash, false)
```
Comment (araskachoi) [inline x/evm/client/rest/rest.go:73]: im having some trouble understanding what an emptyCtx is used for here for `context.Background()`

File: x/evm/genesis.go [hunk 11]
```diff
@@ -20,9 +20,9 @@ func InitGenesis(
 	ctx sdk.Context,
 	k keeper.Keeper,
 	accountKeeper types.AccountKeeper, // nolint: interfacer
+	bankKeeper types.BankKeeper,
 	data GenesisState,
```
Comment (araskachoi) [inline x/evm/genesis.go:24]: is there a reason why we use `GenesisState` here and not `types.GenesisState` like in exportgenesis?

File: x/evm/handler.go [hunk 12]
```diff
@@ -46,117 +49,37 @@ func NewHandler(k *Keeper) sdk.Handler {
...
-func handleMsgEthermint(ctx sdk.Context, k *Keeper, msg types.MsgEthermint) (*sdk.Result, error) {
-	// parse the chainID from a string to a base-10 integer
-	chainIDEpoch, err := ethermint.ParseChainID(ctx.ChainID())
-	if err != nil {
-		return nil, err
-	}
+		switch msg := msg.(type) {
+		case *types.MsgEthereumTx:
+			// execute state transition
+			res, err := k.EthereumTx(sdk.WrapSDKContext(ctx), msg)
```
Comment (araskachoi) [inline x/evm/handler.go:56]: no check for error here

File: x/evm/handler_test.go [hunk 13]
```diff
@@ -70,15 +77,15 @@ func (suite *EvmTestSuite) TestHandleMsgEthereumTx() {
 		{
 			"passed",
```
Comment (araskachoi) [inline x/evm/handler_test.go:78]: what does "passed" test mean?

File: x/evm/handler_test.go [hunk 14]
```diff
@@ -264,31 +215,31 @@ func (suite *EvmTestSuite) TestQueryTxLogs() {
...
-	var txLogs types.QueryETHLogs
-	suite.codec.MustUnmarshalJSON(res, &txLogs)
+	// var txLogs types.QueryETHLogs
+	// suite.codec.MustUnmarshalJSON(res, &txLogs)
 
 	// amino decodes an empty byte array as nil, whereas JSON decodes it as []byte{} causing a discrepancy
-	resultData.Logs[0].Data = []byte{}
-	suite.Require().Equal(txLogs.Logs[0], resultData.Logs[0])
+	// txResponse.TxLogs.Logs[0].Data = []byte{}
+	// suite.Require().Equal(txLogs.Logs[0], txResponse.TxLogs.Logs[0])
```
Comment (araskachoi) [inline x/evm/handler_test.go:242]: should just remove if we are not checking this?

File: x/evm/keeper/abci_test.go [hunk 15]
```diff
@@ -5,37 +5,13 @@ import (
...
-		Header: abci.Header{
-			LastBlockId: abci.BlockID{
-				Hash: []byte("hash"),
-			},
-			Height: 10,
-		},
-	}
-
-	// get the initial consumption
-	initialConsumed := suite.ctx.GasMeter().GasConsumed()
```
Comment (araskachoi) [inline x/evm/keeper/abci_test.go:18]: why don't we check for consumed gas anymore?

File: x/evm/keeper/grpc_query.go [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	if types.IsZeroAddress(req.Address) {
+		return nil, status.Error(
+			codes.InvalidArgument,
+			types.ErrZeroAddress.Error(),
+		)
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
```
Comment (araskachoi) [inline x/evm/keeper/grpc_query.go:32]: looks like this is repeated a few times, make into a helper method?

Comment (araskachoi) [reply x/evm/keeper/grpc_query.go:32]: just a NIT (:

File: x/evm/keeper/grpc_query_test.go [hunk 17]
```diff
@@ -0,0 +1,487 @@
...
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(fmt.Sprintf("Case %s", tc.msg), func() {
+			suite.SetupTest() // reset
+
+			tc.malleate()
+			ctx := sdk.WrapSDKContext(suite.ctx)
+			res, err := suite.queryClient.Code(ctx, req)
```
Comment (araskachoi) [inline x/evm/keeper/grpc_query_test.go:231]: the comment of this method looks like its a copypasta of `Storage(...)` in `x/evm/types/query.pb.proto`

File: x/evm/keeper/grpc_query_test.go [hunk 18]
```diff
@@ -0,0 +1,487 @@
...
+		expBloom []byte
+	)
+
+	testCases := []struct {
+		msg      string
+		malleate func()
+		expPass  bool
+	}{
+		{"bloom bytes not found for height",
+			func() {},
```
Comment (araskachoi) [inline x/evm/keeper/grpc_query_test.go:444]: should we add logic here?
```
bloom := ethtypes.BytesToBloom([]byte("bloom"))
expBloom = bloom.Bytes()
suite.ctx = suite.ctx.WithBlockHeight(1)
``` 
(without `SetBlockBloom(...)`)

File: x/evm/keeper/grpc_query_test.go [hunk 19]
```diff
@@ -0,0 +1,487 @@
...
+
+				suite.Require().Equal(expBloom, res.Bloom)
+			} else {
+				suite.Require().Error(err)
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestQueryParams() {
```
Comment (araskachoi) [inline x/evm/keeper/grpc_query_test.go:480]: should we have a test where there is a non-empty param request?

File: x/evm/keeper/invariants.go [hunk 20]
```diff
@@ -30,18 +30,19 @@ func (k Keeper) BalanceInvariant() sdk.Invariant {
...
 			ethAccount, ok := account.(*ethermint.EthAccount)
 			if !ok {
 				// ignore non EthAccounts
 				return false
 			}
 
 			evmDenom := k.GetParams(ctx).EvmDenom
-			accountBalance := ethAccount.GetCoins().AmountOf(evmDenom)
+
+			accountBalance := k.bankKeeper.GetBalance(ctx, ethAccount.GetAddress(), evmDenom)
```
Comment (araskachoi) [inline x/evm/keeper/invariants.go:42]: question: where in the code does it update both evm & bankkeeper balance when performing a balance transfer?

File: x/evm/keeper/keeper.go [hunk 21]
```diff
@@ -43,7 +43,8 @@ type Keeper struct {
 
 // NewKeeper generates new evm module keeper
 func NewKeeper(
-	cdc *codec.Codec, storeKey sdk.StoreKey, paramSpace params.Subspace, ak types.AccountKeeper,
+	cdc codec.BinaryMarshaler, storeKey sdk.StoreKey, paramSpace paramtypes.Subspace,
```
Comment (araskachoi) [inline x/evm/keeper/keeper.go:46]: can we use the codec.BinaryMarshaler that is set in `paramSpace` instead of `cdc`?

File: x/evm/keeper/keeper_test.go [hunk 22]
```diff
@@ -32,27 +32,31 @@ var (
...
+	address     ethcmn.Address
 }
 
 func (suite *KeeperTestSuite) SetupTest() {
 	checkTx := false
 
 	suite.app = app.Setup(checkTx)
-	suite.ctx = suite.app.BaseApp.NewContext(checkTx, abci.Header{Height: 1, ChainID: "ethermint-3", Time: time.Now().UTC()})
-	suite.querier = keeper.NewQuerier(*suite.app.EvmKeeper)
+	suite.ctx = suite.app.BaseApp.NewContext(checkTx, tmproto.Header{Height: 1, ChainID: "ethermint-3", Time: time.Now().UTC()})
```
Comment (araskachoi) [inline x/evm/keeper/keeper_test.go:45]: is tmproto.Header the same as abci.Header? can't find the old definition of abci.Header and was wondering if it should have proto encoding or is there the possibility that you would use amino to encode the header?

File: x/evm/keeper/keeper_test.go [hunk 23]
```diff
@@ -32,27 +32,31 @@ var (
...
 	suite.address = ethcmn.HexToAddress(addrHex)
 
-	balance := sdk.NewCoins(ethermint.NewPhotonCoin(sdk.ZeroInt()))
+	queryHelper := baseapp.NewQueryServerTestHelper(suite.ctx, suite.app.InterfaceRegistry())
```
Comment (araskachoi) [inline x/evm/keeper/keeper_test.go:48]: this only runs the test on the grpc query but should we also test on rest-server query?

File: x/evm/keeper/keeper_test.go [hunk 24]
```diff
@@ -92,17 +99,19 @@ func (suite *KeeperTestSuite) TestTransactionLogs() {
 		Address:     suite.address,
 		Data:        []byte("log3"),
 		BlockNumber: 10,
+		Topics:      []ethcmn.Hash{},
 	}
 	suite.app.EvmKeeper.AddLog(suite.ctx, log3)
+	log3.Index = 0
 
 	txLogs := suite.app.EvmKeeper.GetAllTxLogs(suite.ctx)
 	suite.Require().Equal(2, len(txLogs))
```
Comment (araskachoi) [inline x/evm/keeper/keeper_test.go:108]: why is this 2? does SetLogs not count as a transaction?

File: x/evm/keeper/grpc_query_test.go (hunk 19 above)
Comment (fedekunze) [reply x/evm/keeper/grpc_query_test.go:480]: there are no args on the request

File: x/evm/keeper/invariants.go (hunk 20 above)
Comment (fedekunze) [reply x/evm/keeper/invariants.go:42]: when the dirty state object is finalized and committed

File: x/evm/keeper/keeper.go (hunk 21 above)
Comment (fedekunze) [reply x/evm/keeper/keeper.go:46]: that is a private codec

File: x/evm/keeper/keeper_test.go (hunk 23 above)
Comment (fedekunze) [reply x/evm/keeper/keeper_test.go:48]: there's no rest server anymore

File: x/evm/module.go [hunk 25]
```diff
@@ -18,71 +22,87 @@ import (
...
-	return types.ModuleCdc.MustMarshalJSON(types.DefaultGenesisState())
+// DefaultGenesis returns default genesis state as raw bytes for the evm
+// module.
+func (AppModuleBasic) DefaultGenesis(cdc codec.JSONMarshaler) json.RawMessage {
+	return cdc.MustMarshalJSON(types.DefaultGenesisState())
 }
 
 // ValidateGenesis is the validation check of the Genesis
-func (AppModuleBasic) ValidateGenesis(bz json.RawMessage) error {
+func (AppModuleBasic) ValidateGenesis(cdc codec.JSONMarshaler, _ client.TxEncodingConfig, bz json.RawMessage) error {
```
Comment (araskachoi) [inline x/evm/module.go:49]: is _ client.TxEncodingConfig added as an arg to fulfill the interface?

File: x/evm/module.go [hunk 26]
```diff
@@ -18,71 +22,87 @@ import (
...
 }
 
+// GetQueryCmd returns no root query command for the evm module.
+func (AppModuleBasic) GetQueryCmd() *cobra.Command {
+	return cli.GetQueryCmd()
+}
+
+// RegisterInterfaces registers interfaces and implementations of the evm module.
+func (AppModuleBasic) RegisterInterfaces(registry codectypes.InterfaceRegistry) {
+	types.RegisterInterfaces(registry)
```
Comment (araskachoi) [inline x/evm/module.go:82]: how come this method does a `registry.RegisterImplementations(...)` twice inside the `types.RegisterInterfaces(registry)`?

**Review by araskachoi — COMMENTED** [review]
have some comments if you can respond to them, but looks good for the most part!

File: x/evm/module.go (hunk 26 above)
Comment (fedekunze) [reply x/evm/module.go:82]: so you register the concrete implementation with the interface. The interfaces are different
